### PR TITLE
[InstCombine] Fold fmod to frem if we know it does not set errno.

### PR DESCRIPTION
**Pull request by davemgreen**
fmod will be folded to frem in clang under -fno-math-errno and can be constant
folded in llvm if the operands are known. It can be relatively common to have
fp code that handles special values before doing some calculation:
```
if (isnan(f))
  return handlenan;
if (isinf(f))
  return handleinf;
..
fmod(f, 2.0)
```

This patch enables the folding of fmod to frem in instcombine if the first
parameter is not inf and the second is not zero. Other combinations do not set
errno.

### CODE REVIEW

**Review by nikic — COMMENTED** [review]
This should be in SimplifyLibCalls.

**Review by nikic — COMMENTED** [review]
This looks fine to me, but someone more familiar with FP should take a look at this...

File: llvm/lib/Transforms/Utils/SimplifyLibCalls.cpp [hunk 1]
```diff
@@ -2278,8 +2278,8 @@ Value *LibCallSimplifier::replacePowWithSqrt(CallInst *Pow, IRBuilderBase &B) {
   // pow(-Inf, 0.5) is optionally required to have a result of +Inf (not setting
   // errno), but sqrt(-Inf) is required by various standards to set errno.
   if (!Pow->doesNotAccessMemory() && !Pow->hasNoInfs() &&
-      !isKnownNeverInfinity(Base, 0,
-                            SimplifyQuery(DL, TLI, /*DT=*/nullptr, AC, Pow)))
+      !isKnownNeverInfinity(
+          Base, 0, SimplifyQuery(DL, TLI, DT, AC, Pow, true, true, DC)))
```
Comment (arsenm) [inline llvm/lib/Transforms/Utils/SimplifyLibCalls.cpp:2282]: Passing in the new analyses are a separate patch

File: llvm/lib/Transforms/Utils/SimplifyLibCalls.cpp [hunk 2]
```diff
@@ -2796,6 +2796,25 @@ Value *LibCallSimplifier::optimizeSqrt(CallInst *CI, IRBuilderBase &B) {
...
+Value *LibCallSimplifier::optimizeFMod(CallInst *CI, IRBuilderBase &B) {
+  SimplifyQuery SQ(DL, TLI, DT, AC, CI, true, true, DC);
+
+  // fmod(x,y) can set errno if y == 0 or x == +/-inf.
+  KnownFPClass Known0 = computeKnownFPClass(CI->getOperand(0), fcInf,
+                                            /*Depth=*/0, SQ);
+  if (Known0.isKnownNeverInfinity()) {
+    KnownFPClass Known1 = computeKnownFPClass(CI->getOperand(1), fcZero,
+                                              /*Depth=*/0, SQ);
+    if (Known1.isKnownNeverZero()) {
```
Comment (arsenm) [inline llvm/lib/Transforms/Utils/SimplifyLibCalls.cpp:2808]: This is wrong if DAZ is enabled, this needs to account of the denormal mode. Can you add a negative test with "denormal-fp-math"="preserve-sign,preserve-sign" and/or ="dynamic,dynamic"? This should be checking isKnownNeverLogicalZero

File: llvm/lib/Transforms/Utils/SimplifyLibCalls.cpp (hunk 1 above)
Comment (davemgreen) [reply llvm/lib/Transforms/Utils/SimplifyLibCalls.cpp:2282]: I'll rebase over #108446 once it goes in, probably tomorrow morning.

File: llvm/lib/Transforms/Utils/SimplifyLibCalls.cpp (hunk 2 above)
Comment (davemgreen) [reply llvm/lib/Transforms/Utils/SimplifyLibCalls.cpp]: Thanks - my testing (https://gist.github.com/davemgreen/f6b084618c24db2b0996cc618cd8ca05) was on a normal linux system. There is a test in the end of ad3ad15229cc65874eae173137e47b5c817d77ea, and I've changed it to isKnownNeverLogicalZero.

File: llvm/lib/Transforms/Utils/SimplifyLibCalls.cpp [hunk 3]
```diff
@@ -2796,6 +2796,27 @@ Value *LibCallSimplifier::optimizeSqrt(CallInst *CI, IRBuilderBase &B) {
...
 
+Value *LibCallSimplifier::optimizeFMod(CallInst *CI, IRBuilderBase &B) {
+  SimplifyQuery SQ(DL, TLI, DT, AC, CI, true, true, DC);
+
+  // fmod(x,y) can set errno if y == 0 or x == +/-inf. Otherwise we can change
+  // it to a frem instruction.
+  KnownFPClass Known0 = computeKnownFPClass(CI->getOperand(0), fcInf,
+                                            /*Depth=*/0, SQ);
+  if (Known0.isKnownNeverInfinity()) {
+    KnownFPClass Known1 = computeKnownFPClass(CI->getOperand(1), fcZero,
```
Comment (arsenm) [inline llvm/lib/Transforms/Utils/SimplifyLibCalls.cpp:2807]: fcZero | fcSubnormal

File: llvm/lib/Transforms/Utils/SimplifyLibCalls.cpp [hunk 4]
```diff
@@ -2796,6 +2796,27 @@ Value *LibCallSimplifier::optimizeSqrt(CallInst *CI, IRBuilderBase &B) {
   return copyFlags(*CI, FabsCall);
 }
 
+Value *LibCallSimplifier::optimizeFMod(CallInst *CI, IRBuilderBase &B) {
+  SimplifyQuery SQ(DL, TLI, DT, AC, CI, true, true, DC);
+
+  // fmod(x,y) can set errno if y == 0 or x == +/-inf. Otherwise we can change
+  // it to a frem instruction.
+  KnownFPClass Known0 = computeKnownFPClass(CI->getOperand(0), fcInf,
```
Comment (arsenm) [inline llvm/lib/Transforms/Utils/SimplifyLibCalls.cpp:2804]: You can also just do this if the instruction has the nnan flag

**Review by arsenm — APPROVED** [review]
LGTM, but this is ignoring the flags directly on the instruction. If I read this correctly, you can just do this for any nnan call

**Comment by davemgreen** [comment]
I've had a go at making it work of nnan instructions too. Let me know if this still looks OK.

**Review by arsenm — APPROVED** [review]

**Pull request merged**